design/design-doc-composition.md

```diff
@@ -0,0 +1,1079 @@
+only uses AWS. If this organisation values Crossplane's separation of concerns
+but does not need its portability there is no reason that its application
+operators should be limited to resource claims that are constrained by
+supporting all possible providers.
+
+## Goals
+
+The goal of this document is to propose a simple, consistent, Kubernetes native
+model for composing applications and their infrastructure needs.
+
+TODO(negz): Elaborate on the below goals.
+
+* Nested composition
+* User defined abstractions
+* Shareable abstractions
+* Composable infrastructure
+* Discoverable abstraction to composition relationships
+* Compositions configured inside API boundary
+
+## Use Cases
+
+The proposal put forward by this document is intended to address a handful of
+use cases around application and infrastructure provisioning.
+
+### Application Composition
+
+An _application developer_ wishes to simplify and standardise the work of the
+_application operators_ who will deploy and operate their application. They wish
+to encode the deployment and infrastructure needs of their application, and to
+expose a subset of the configuration of said deployment and infrastructure needs
+to the _application operators_.
+
+For example, the _(application) developers_ of Wordpress wish to encode the fact
+that each deployment of Wordpress requires one or more web servers running
+wordpress, an SQL database, and some kind of network ingress (e.g. a load
+balancer) to get requests to the web servers.
+
+Wordpress’s _application developers_ wish to allow _application operators_ to
+configure only how much storage their SQL database has, whether the database
+uses SSD or spinning disk, and how replicated the web server will be.
+
+All other configuration details should be selected from a “class” or “profile”of
+application. This class may decide whether the web server is containerized or a
+VM, whether the SQL database is PostgreSQL or MySQL, etc.
+
+The input to this example is a `kind: Wordpress` custom resource. The outputs
+depend on the class of application, but could include `kind: Deployment`, and
+`kind: MySQLInstance`, where `kind: MySQLInstance` is a namespaced composite
+infrastructure resource.
+
+GitLab is a more advanced example of an application composition, in which the
+GitLab application is composed in turn of many smaller composite applications
+each roughly as complex as this Wordpress example.
+
+In this example:
+
+* The _schema of the input_ (`kind: Wordpress`) is defined by an _app
+  developer_.
+* The _configuration of the input_ is defined by an _application operator_.
+* The _schema of the output_ composed resources (e.g. `kind: MySQLInstance`) may
+  be defined by either an _application developer or an infrastructure operator_.
+* The _configuration of the output_ composed resources is defined by an
+  _application developer_.
+
+### Infrastructure Composition
```

Comment:
Just an observation. This model assumes that there are infrastructure provider/operator and application operator/developer. Not sure how often this actually happens in real world.

Reply:
Yeah, I think all four(ish) of these personas are a venn diagram with quite a lot of overlap in the real world. The challenge is to create a platform that accommodates the spectrum from one engineer playing all roles (a tiny startup) through to distinct engineers in each role (a larger, perhaps slightly old fashioned org). This challenge exists for any platform that acknowledges a separation of concerns, including Crossplane and any OAM runtime.